ndonnx/_utility.py

```diff
@@ -45,7 +45,16 @@ def _promote_with_none(*args: Array | npt.ArrayLike) -> list[Array | None]:
                 signed_integer = True
         elif isinstance(arg, np.ndarray):
             arr_or_none.append(asarray(arg))
-        elif isinstance(arg, (int, float, str, np.generic)):
+        elif isinstance(arg, float):
```

Comment:
This currently works because some `np.generic`s such as `np.float64` are also instances of `float`. Is this deliberate? If so, how would we end up with a `float16` further below? How about we give up on the idea of implicitly finding the smallest dtype in favor of some simpler logic here?

Reply:
I meant to delete the line from before with the `float16` handling as it is now dead code, thanks for point it out.

I don't think you can give up on this, it is tested quite a lot in the array-api-standard tests and the type promotion DAG seems to suggest you go to `float32` if it is valid to do so.

Reply:
Thanks for the link! The section on dealing with Python scalars seems very relevant and rather easy to follow. Furthermore, there is this statement which seem to be a clear argument against using `numpy.min_scalar_type`: 
> Only the types of participating arrays matter, not their values

Reply:
It says the types of the **arrays** matter. For mixing arrays and [scalars](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#mixing-arrays-with-python-scalars) it says that the scalar adopts the dtype of the array. This ends up happening here by taking the minimum dtype and promoting the arguments (and also works for `result_type`).

Reply:
> For mixing arrays and [scalars](https://data-apis.org/array-api/latest/API_specification/type_promotion.html#mixing-arrays-with-python-scalars) it says that the scalar adopts the dtype of the array. This ends up happening here by taking the minimum dtype the scalar can represent and promoting the arguments

What you describe is not "adopting the data type of the array", though. Instead, the algorithm you describe may result in a data type depending on the value of the Python scalar, doesn't it. Consider a value such as `1.1` which is more precisely represented in a `float64` than a `float32`. Further, consider an array of data type `float32`. As I read the standard, the promotion of the two should result in `float32` not `float64`. This is also what NumPy 2 is doing:

```python
import numpy as np
assert np.__version__ == "2.0.0"

val = 1.1
assert np.float32(val) != np.float64(val)

arr = np.array(1.0, np.float32) + val
assert arr.dtype == np.float32
```

Ultimately, I suppose our problem may boil down to the fact that type promotion is only defined by the standard if at least one argument is actually a standard `Array`.

Reply:
Fair enough. It's interesting that the precision of the python scalar ends up having no relevance (and so you can lose precision like in the example), even though it is not an array yet.

Reply:
I should have looked into this a bit more. I'm not sure this is entirely consistent and it leads to some fairly odd situations now:

```
(Pdb) np.asarray([1, 2]) + 2.32
array([3.32, 4.32])
(Pdb) ndx.asarray([1, 2]) + 2.32
Array([3 4], dtype=Int64)
```

Reply:
Oh yes, that aint good! I suppose we are still not following the letter of the standard, though:

> ...
>   - a Python int or float for real-valued floating-point array data types.
>
> ...
> Provided the above requirements are met, the expected behavior is equivalent to:
>  1. Convert the scalar to zero-dimensional array with the same data type as that of the array used in the expression.
>  2. Execute the operation for array <op> 0-D array (or 0-D array <op> array if scalar was the left-hand argument).

Reply:
Oh no, that ain't good! Let's discuss in the follow up PR.